common/go-utils/src/vrs.rs

```diff
@@ -1,76 +1,74 @@
+use std::fmt::{self};
 
 /// `Requirement` is a wrapper around `semver::Requirement` that adds
-/// - `Deserialize` and `Serialize` traits
 /// - Ability to parse go-flavored requirements
 ///
 /// The derived `Default` implementation creates a wildcard version `Requirement`.
-#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq)]
-#[serde(try_from = "String", into = "String")]
-pub struct Requirement(semver::VersionReq);
+#[derive(Debug, Default, PartialEq, Clone)]
+pub struct GoRequirement(semver::VersionReq);
 
 #[derive(thiserror::Error, Debug)]
 #[error("Couldn't parse Go version requirement: {0}")]
 pub struct RequirementParseError(#[from] semver::Error);
 
-impl Requirement {
+pub trait Version {}
+
+impl Version for GoVersion {}
+pub trait VersionRequirement<T> {
+    fn satisfies(&self, version: &T) -> bool;
+
     /// Parses a semver requirement `&str` as a `Requirement`.
     ///
     /// # Examples
     ///
     /// ```
-    /// let req = heroku_go_utils::vrs::Requirement::parse("~1.0").unwrap();
+    /// use heroku_go_utils::vrs::VersionRequirement;
+    /// let req = heroku_go_utils::vrs::GoRequirement::parse("~1.0").unwrap();
     /// ```
     ///
     /// # Errors
     /// Invalid semver requirement `&str` like ">< 1.0", ".1.0", "!=4", etc.
     /// will return an error.
-    pub fn parse(input: &str) -> Result<Self, RequirementParseError> {
-        Ok(semver::VersionReq::parse(input).map(Self)?)
+    fn parse(input: &str) -> Result<Self, RequirementParseError>
+    where
+        Self: std::marker::Sized;
+}
+
+impl VersionRequirement<GoVersion> for GoRequirement {
+    fn satisfies<'a>(&self, version: &GoVersion) -> bool {
+        self.0.matches(&version.0)
     }
 
     /// Parses a go version requirement `&str` as a `Requirement`
     ///
     /// # Examples
     ///
     /// ```
-    /// let req = heroku_go_utils::vrs::Requirement::parse_go("go1.0").unwrap();
+    /// use heroku_go_utils::vrs::VersionRequirement;
+    /// let req = heroku_go_utils::vrs::GoRequirement::parse("go1.0").unwrap();
     /// ```
     ///
     /// # Errors
     /// Invalid semver requirement `&str` like ">< 1.0", ".1.0", "!=4", etc.
     /// will return an error.
-    pub fn parse_go(go_req: &str) -> Result<Self, RequirementParseError> {
-        let stripped_req = go_req
+    fn parse(input: &str) -> Result<Self, RequirementParseError> {
+        input
             .strip_prefix("go")
-            .map_or(Cow::Borrowed(go_req), |req| Cow::Owned(format!("={req}")));
-        Self::parse(&stripped_req)
-    }
-
-    /// Determines if a `&Version` satisfies a `Requirement`
-    #[must_use]
-    pub fn satisfies(&self, version: &Version) -> bool {
-        self.0.matches(&version.0)
-    }
-}
-
-impl TryFrom<String> for Requirement {
-    type Error = RequirementParseError;
-    fn try_from(val: String) -> Result<Self, Self::Error> {
-        Requirement::parse(&val)
+            .map_or(parse_semver_requirement(input), |req| {
```

Comment:
Something I find confusing about the tests around this area is the following:
- A requirement of `go1` is parsed to `=1`
- A requirement of `1` is parsed to `^1`

Why does the presence of the `go` prefix seem to produce two different results?

Reply:
It's mostly to support the classic buildpack mechanism where users can specify `// +heroku goVersion go1.11`. The resolution strategy there is not semver, it's something specific to go and our classic go buildpack, which is roughly equivalent to `=1.11` in semver. Users can also use standard semver requirements, like `// +heroku goVersion ~1.16` in the CNB, since we have strong semver parsing built in. The parsing of `1` into `^1` is just how the semver crate works by default.

Reply:
Okay, that makes sense.  I think part of my confusion was also related to how this part of the function is written:

```rust
input
  .strip_prefix("go")
  .map_or(parse_semver_requirement(input), |req| {
    parse_semver_requirement(format!("={req}").as_str())
  })
```

I find myself backtracking on this code which seems straightforward at first glance but requires some careful unpacking to follow.

@runesoerensen Could I suggest the following edit?

```rust
if let Some(input_without_prefix) = input.strip_prefix("go") {
  parse_semver_requirement(format!("={input_without_prefix}").as_str())
} else {
  parse_semver_requirement(input)
}
```

Reply:
I ended up dropping the `GoRequirement` abstraction altogether, and simply moved this logic to a function that returns a `semver::VersionReq` instead: https://github.com/heroku/buildpacks-go/pull/245/files#diff-560d0a7bad62e4a300eba035429762f445bb85bb6ab9bc20f42b02c22545fb3cR25-R31

I think the new implementation makes it a bit clearer what's happening here?